server/fleet/sessions.go

```diff
@@ -49,3 +71,104 @@ type Session struct {
+// or member of teams).
+type SSORolesInfo struct {
+	// Global holds the role for the Global domain.
+	Global *string
+	// Teams holds the roles for teams.
+	Teams []TeamRole
+}
+
+// TeamRole holds a user's role on a Team.
+type TeamRole struct {
+	// ID is the unique identifier of the team.
+	ID uint
+	// Role is the role of the user in the team.
+	Role string
+}
+
+func (s SSORolesInfo) verify() error {
+	if s.Global != nil && len(s.Teams) > 0 {
+		return errors.New("cannot set both global and team roles")
+	}
+	// Check for duplicate entries for the same team.
+	// This is just in case some IdP allows duplicating attributes.
+	teamSet := make(map[uint]struct{})
+	for _, teamRole := range s.Teams {
+		if _, ok := teamSet[teamRole.ID]; ok {
+			return fmt.Errorf("duplicate team entry: %d", teamRole.ID)
+		}
+		teamSet[teamRole.ID] = struct{}{}
+	}
+	return nil
+}
+
+func (s SSORolesInfo) isEmpty() bool {
+	return s.Global == nil && len(s.Teams) == 0
+}
+
+const (
+	globalUserRoleSSOAttrName     = "FLEET_JIT_USER_ROLE_GLOBAL"
+	teamUserRoleSSOAttrNamePrefix = "FLEET_JIT_USER_ROLE_TEAM_"
+)
+
+// RolesFromSSOAttributes loads Global and Team roles from SAML custom attributes.
+//   - Custom attribute `FLEET_JIT_USER_ROLE_GLOBAL` is used for setting global role.
+//   - Custom attributes of the form `FLEET_JIT_USER_ROLE_TEAM_<TEAM_ID>` are used
+//     for setting role for a team with ID <TEAM_ID>.
+//
+// For both attributes currently supported values are `admin`, `maintainer` and `observer`
+func RolesFromSSOAttributes(attributes []SAMLAttribute) (SSORolesInfo, error) {
+	ssoRolesInfo := SSORolesInfo{}
+	for _, attribute := range attributes {
+		switch {
+		case attribute.Name == globalUserRoleSSOAttrName:
+			role, err := parseRole(attribute.Values)
+			if err != nil {
+				return SSORolesInfo{}, fmt.Errorf("parse global role: %w", err)
+			}
+			ssoRolesInfo.Global = ptr.String(role)
+		case strings.HasPrefix(attribute.Name, teamUserRoleSSOAttrNamePrefix):
+			teamIDSuffix := strings.TrimPrefix(attribute.Name, teamUserRoleSSOAttrNamePrefix)
+			teamID, err := strconv.ParseUint(teamIDSuffix, 10, 32)
+			if err != nil {
+				return SSORolesInfo{}, fmt.Errorf("parse team ID: %w", err)
+			}
+			teamRole, err := parseRole(attribute.Values)
+			if err != nil {
+				return SSORolesInfo{}, fmt.Errorf("parse team role: %w", err)
+			}
+			ssoRolesInfo.Teams = append(ssoRolesInfo.Teams, TeamRole{
+				ID:   uint(teamID),
+				Role: teamRole,
+			})
+		default:
+			continue
+		}
+	}
+	if err := ssoRolesInfo.verify(); err != nil {
+		return SSORolesInfo{}, err
+	}
+	if ssoRolesInfo.isEmpty() {
+		// When the configuration is not set, the default is to
+		// make the user a global observer.
+		return SSORolesInfo{Global: ptr.String(RoleObserver)}, nil
+	}
+	return ssoRolesInfo, nil
+}
+
+func parseRole(values []SAMLAttributeValue) (string, error) {
+	if len(values) == 0 {
+		return "", errors.New("empty role")
+	}
+	// Using last value by default.
```

Comment:
Why are we using last value?

Reply:
There's no requirement around multi-value attributes, so we make a decision here.
Which reminds me we should document it.

Reply:
Good catch! Added in https://github.com/fleetdm/fleet/commit/9bff202bc98ab707a7ab4eb2f8ef86eefb1c40f4.